File: lizzie-mobile/shared/src/commonMain/kotlin/com/lizzie/rules/SGFParser.kt
package com.lizzie.rules

/**
 * SGF (Smart Game Format) parser and serializer for Go.
 *
 * Ported from Lizzie's SGFParser.java.
 * Handles standard FF[4] SGF files with Go game trees.
 */
object SGFParser {

    /**
     * Parse SGF text into a game tree.
     * Returns the root properties and the ordered list of move nodes.
     */
    fun parseSgf(sgfContent: String): ParsedSgfGame {
        val cleaned = sgfContent.trim()
        if (!cleaned.startsWith("(") || !cleaned.endsWith(")")) {
            throw IllegalArgumentException("Invalid SGF: missing wrapping parentheses")
        }

        val gameInfo = GameInfo()
        val moves = mutableListOf<ParsedMove>()
        var komi = 6.5
        var boardSize = 19
        var foundRoot = false

        // Parse tokens: ( ) ; and key[value] pairs
        val tokens = tokenize(cleaned)

        var i = 0
        while (i < tokens.size) {
            when (tokens[i]) {
                "(" -> { /* start of game tree — skip */ i++ }
                ")" -> { /* end — skip */ i++ }
                ";" -> {
                    i++ // read properties until next ";" or "(" or ")"
                    val props = mutableMapOf<String, String>()
                    while (i < tokens.size && tokens[i] !in setOf(";", "(", ")")) {
                        val key = tokens[i]
                        i++
                        if (i < tokens.size && tokens[i].startsWith("[")) {
                            val value = tokens[i].trimStart('[').trimEnd(']')
                            props[key] = value
                            i++
                        }
                    }

                    if (!foundRoot) {
                        // Root node — extract game info
                        foundRoot = true
                        boardSize = props["SZ"]?.toIntOrNull() ?: 19
                        komi = props["KM"]?.toDoubleOrNull() ?: 6.5
                        gameInfo.copy(
                            playerBlack = props["PB"] ?: "",
                            playerWhite = props["PW"] ?: "",
                            komi = komi,
                            handicap = props["HA"]?.toIntOrNull() ?: 0,
                            ruleSet = props["RU"] ?: "jp",
                            gameName = props["GN"] ?: "",
                            gameDate = props["DT"] ?: "",
                            result = props["RE"] ?: "",
                        )

                        // Handle handicap setup stones (AB)
                        val abStones = parseSgfStones(props["AB"] ?: "")
                        for (stone in abStones) {
                            moves.add(ParsedMove(stone.first, stone.second, Stone.BLACK, isSetup = true))
                        }
                        val awStones = parseSgfStones(props["AW"] ?: "")
                        for (stone in awStones) {
                            moves.add(ParsedMove(stone.first, stone.second, Stone.WHITE, isSetup = true))
                        }
                    } else {
                        // Move node
                        val b = props["B"]?.takeIf { it.isNotBlank() }
                        val w = props["W"]?.takeIf { it.isNotBlank() }

                        if (b != null) {
                            val coord = sgfCoordToBoard(b, boardSize)
                            if (coord != null) {
                                moves.add(ParsedMove(coord.first, coord.second, Stone.BLACK, isPass = false))
                            } else {
                                moves.add(ParsedMove(0, 0, Stone.BLACK, isPass = true))
                            }
                        } else if (w != null) {
                            val coord = sgfCoordToBoard(w, boardSize)
                            if (coord != null) {
                                moves.add(ParsedMove(coord.first, coord.second, Stone.WHITE, isPass = false))
                            } else {
                                moves.add(ParsedMove(0, 0, Stone.WHITE, isPass = true))
                            }
                        }

                        // Comment
                        val comment = props["C"]
                        if (comment != null && moves.isNotEmpty()) {
                            val lastIdx = moves.lastIndex
                            moves[lastIdx] = moves[lastIdx].copy(comment = comment)
                        }
                    }

                    // Skip nested subtrees for now (basic parser)
                    // Lizzie's original parser handles variations as different branches
                }
                else -> i++ // unknown token
            }
        }

        return ParsedSgfGame(gameInfo, boardSize, moves)
    }

    /**
     * Serialize a game to SGF string.
     */
    fun saveSgf(
        history: BoardHistoryList,
        gameInfo: GameInfo,
        width: Int = 19,
        height: Int = 19
    ): String {
        val sb = StringBuilder()
        sb.append("(;FF[4]GM[1]SZ[$width]")
        sb.append("KM[${gameInfo.komi}]")
        if (gameInfo.playerBlack.isNotEmpty()) sb.append("PB[${escapeSgf(gameInfo.playerBlack)}]")
        if (gameInfo.playerWhite.isNotEmpty()) sb.append("PW[${escapeSgf(gameInfo.playerWhite)}]")
        if (gameInfo.result.isNotEmpty()) sb.append("RE[${escapeSgf(gameInfo.result)}]")
        if (gameInfo.gameName.isNotEmpty()) sb.append("GN[${escapeSgf(gameInfo.gameName)}]")
        if (gameInfo.gameDate.isNotEmpty()) sb.append("DT[${escapeSgf(gameInfo.gameDate)}]")
        if (gameInfo.handicap > 0) sb.append("HA[${gameInfo.handicap}]")
        if (gameInfo.ruleSet.isNotEmpty()) sb.append("RU[${escapeSgf(gameInfo.ruleSet)}]")

        // Serialize move tree
        serializeNode(sb, history.root(), width, height)

        sb.append(")")
        return sb.toString()
    }

    private fun serializeNode(sb: StringBuilder, node: BoardHistoryNode?, width: Int, height: Int) {
        if (node == null) return
        val data = node.data
        val isRoot = node.previous() == null

        if (isRoot) {
            // Root node already written its properties above
            val variations = node.getVariations()
            for (child in variations) {
                serializeNode(sb, child, width, height)
            }
            return
        }

        sb.append(";")
        val color = data.lastMoveColor
        val coord = data.lastMove
        if (coord != null) {
            val sgfCoord = boardCoordToSgf(coord.first, coord.second, width, height)
            if (color == Stone.BLACK) {
                sb.append("B[$sgfCoord]")
            } else {
                sb.append("W[$sgfCoord]")
            }
        } else {
            // Pass
            if (color == Stone.BLACK) {
                sb.append("B[]")
            } else {
                sb.append("W[]")
            }
        }

        if (data.comment.isNotEmpty()) {
            sb.append("C[${escapeSgf(data.comment)}]")
        }

        // Check for variations (multiple children)
        val variations = node.getVariations()
        if (variations.size > 1) {
            // First child in main line, rest in parentheses
            val first = variations.first()
            serializeNode(sb, first, width, height)

            for (i in 1 until variations.size) {
                sb.append("(")
                serializeNode(sb, variations[i], width, height)
                sb.append(")")
            }
        } else if (variations.size == 1) {
            serializeNode(sb, variations.first(), width, height)
        }
    }

    // ---- SGF coordinate conversion ----

    /** SGF coordinates use "ab" for the top-left, with letters a=0, b=1, ... z=25. */
    private const val SGF_ALPHABET = "abcdefghijklmnopqrstuvwxyz"

    fun sgfCoordToBoard(sgfCoord: String, boardSize: Int = 19): Pair<Int, Int>? {
        if (sgfCoord.length < 2) return null
        val x = SGF_ALPHABET.indexOf(sgfCoord[0].lowercaseChar())
        val y = SGF_ALPHABET.indexOf(sgfCoord[1].lowercaseChar())
        if (x < 0 || y < 0 || x >= boardSize || y >= boardSize) return null
        return Pair(x, boardSize - 1 - y) // SGF y=0 is top, our y=0 is bottom
    }

    fun boardCoordToSgf(x: Int, y: Int, width: Int = 19, height: Int = 19): String {
        if (x < 0 || y < 0 || x >= width || y >= height) return ""
        return "${SGF_ALPHABET[x]}${SGF_ALPHABET[height - 1 - y]}"
    }

    fun parseSgfStones(value: String): List<Pair<Int, Int>> {
        val coords = mutableListOf<Pair<Int, Int>>()
        val cleaned = value.replace("\\s".toRegex(), "")
        // Pairs of two characters (e.g., "ab" means position a,b)
        var i = 0
        while (i + 1 < cleaned.length) {
            val x = SGF_ALPHABET.indexOf(cleaned[i].lowercaseChar())
            val y = SGF_ALPHABET.indexOf(cleaned[i + 1].lowercaseChar())
            if (x >= 0 && y >= 0) {
                coords.add(Pair(x, y))
            }
            i += 2
        }
        return coords
    }

    // ---- SGF property helpers ----

    fun addProperty(properties: MutableMap<String, String>, key: String, value: String) {
        when (key) {
            "B", "W", "AB", "AW", "AE" -> {
                // Move properties are handled separately
            }
            else -> {
                val existing = properties[key]
                if (existing != null) {
                    properties[key] = "$existing:$value"
                } else {
                    properties[key] = value
                }
            }
        }
    }

    fun addProperties(properties: MutableMap<String, String>, addProps: Map<String, String>) {
        for ((key, value) in addProps) {
            addProperty(properties, key, value)
        }
    }

    fun getOrDefault(properties: Map<String, String>, key: String, default: String): String {
        return properties[key] ?: default
    }

    fun propertiesString(properties: Map<String, String>): String {
        return properties.entries.joinToString(" ") { (key, value) ->
            "$key[$value]"
        }
    }

    // ---- Tokens ----

    private fun tokenize(sgf: String): List<String> {
        val tokens = mutableListOf<String>()
        var i = 0
        while (i < sgf.length) {
            when (val c = sgf[i]) {
                '(', ')' -> {
                    tokens.add(c.toString())
                    i++
                }
                ';' -> {
                    tokens.add(";")
                    i++
                }
                '\\' -> {
                    // Escape — skip next character
                    i += 2
                }
                '[' -> {
                    // Read value until matching ]
                    val start = i
                    i++
                    val value = StringBuilder()
                    while (i < sgf.length && sgf[i] != ']') {
                        if (sgf[i] == '\\' && i + 1 < sgf.length) {
                            value.append(sgf[i + 1])
                            i += 2
                        } else {
                            value.append(sgf[i])
                            i++
                        }
                    }
                    if (i < sgf.length) i++ // skip ]
                    tokens.add("[$value]")
                }
                else -> {
                    if (!c.isWhitespace()) {
                        val key = StringBuilder()
                        while (i < sgf.length && sgf[i] !in ";([]) \t\r\n" && !sgf[i].isWhitespace()) {
                            key.append(sgf[i])
                            i++
                        }
                        tokens.add(key.toString())
                    } else {
                        i++
                    }
                }
            }
        }
        return tokens
    }

    private fun escapeSgf(s: String): String {
        return s.replace("\\", "\\\\")
            .replace("]", "\\]")
            .replace("\n", "\\n")
    }
}

data class ParsedMove(
    val x: Int,
    val y: Int,
    val color: Stone,
    val isPass: Boolean = false,
    val isSetup: Boolean = false,
    val comment: String = "",
)

data class ParsedSgfGame(
    val gameInfo: GameInfo,
    val boardSize: Int,
    val moves: List<ParsedMove>,
)

File: lizzie_mobile/lib/go/sgf_parser.dart
import 'board_data.dart';
import 'board_history_list.dart';
import 'board_history_node.dart';
import 'coords.dart';
import 'game_info.dart';
import 'stone.dart';
import 'zobrist.dart';

/// SGF (Smart Game Format) parser and serializer for Go game records.
///
/// Ported from Lizzie's `SGFParser.java` (1040 lines).
///
/// Supports:
/// - Standard SGF properties: B, W, SZ, KM, HA, AB, AW, PB, PW, RE, C
/// - Nested variations via `(...)` branching
/// - Both standard and MultiGo-style branching
/// - Lizzie-private LZ tag (best-move persistence)
/// - Fox SGF komi correction
class SgfParser {
  /// Load an SGF string into a [BoardHistoryList] with its metadata.
  /// Returns a new BoardHistoryList independent of any live board.
  static (BoardHistoryList, GameInfo) parse(
    String sgf, {
    int defaultSize = 19,
  }) {
    // Strip whitespace and normalize.
    final cleaned = sgf.replaceAll(RegExp(r'\s+'), ' ').trim();

    // Find the root node data block inside the outermost parentheses.
    final rootMatch = RegExp(r'\(([\s\S]*)\)').firstMatch(cleaned);
    if (rootMatch == null) {
      // No parentheses — try parsing as a single node.
      final data = BoardData.empty(width: defaultSize, height: defaultSize);
      return (BoardHistoryList(data), GameInfo());
    }

    final rootContent = rootMatch.group(1)!.trim();
    final (history, gameInfo) = _parseNodes(
      rootContent,
      defaultSize,
      GameInfo(),
    );
    return (history, gameInfo);
  }

  /// Parse a sequence of nodes from the SGF content.
  static (BoardHistoryList, GameInfo) _parseNodes(
    String content,
    int defaultSize,
    GameInfo baseInfo,
  ) {
    // Extract root node properties.
    final rootProps = <String, String>{};
    final rootNodeMatch = RegExp(r'^;([^;()]*)').firstMatch(content);
    String rest;
    if (rootNodeMatch != null) {
      _extractProperties(rootNodeMatch.group(1)!, rootProps);
      rest = content.substring(rootNodeMatch.end).trim();
    } else {
      rest = content.trim();
    }

    // Resolve board size.
    final szStr = rootProps['SZ'] ?? rootProps['sz'] ?? '$defaultSize';
    final dims = szStr.split(':');
    final width = int.tryParse(dims[0]) ?? defaultSize;
    final height = int.tryParse(dims.length > 1 ? dims[1] : dims[0]) ?? width;

    final history = BoardHistoryList(
      BoardData.empty(width: width, height: height),
    );
    final zobrist = Zobrist.init(width, height, 0);

    // Handle setup stones (AB/AW) at root.
    final abList = rootProps['AB'] ?? rootProps['ab'] ?? '';
    final awList = rootProps['AW'] ?? rootProps['aw'] ?? '';
    if (abList.isNotEmpty || awList.isNotEmpty) {
      _applySetupStones(history, abList, awList, zobrist);
    }

    // Extract game metadata from root properties.
    final komi =
        double.tryParse(rootProps['KM'] ?? rootProps['km'] ?? '') ??
        baseInfo.komi;
    final handicap =
        int.tryParse(rootProps['HA'] ?? rootProps['ha'] ?? '') ??
        baseInfo.handicap;
    final playerBlack =
        rootProps['PB'] ?? rootProps['pb'] ?? baseInfo.playerBlack;
    final playerWhite =
        rootProps['PW'] ?? rootProps['pw'] ?? baseInfo.playerWhite;
    final result = rootProps['RE'] ?? rootProps['re'] ?? baseInfo.result;
    final gameName = rootProps['GN'] ?? rootProps['gn'] ?? baseInfo.gameName;
    final gameDate = rootProps['DT'] ?? rootProps['dt'] ?? baseInfo.gameDate;
    baseInfo = GameInfo(
      komi: komi,
      handicap: handicap,
      playerBlack: playerBlack,
      playerWhite: playerWhite,
      result: result,
      gameName: gameName,
      gameDate: gameDate,
    );

    // Now parse the move sequence and variations from the remaining content.
    if (rest.isNotEmpty) {
      _parseBranchContent(rest, history, width, height, zobrist);
    }

    // Rewind to start.
    history.toStart();
    return (history, baseInfo);
  }

  /// Apply AB/AW setup stones to the history root.
  static void _applySetupStones(
    BoardHistoryList history,
    String abList,
    String awList,
    Zobrist zobrist,
  ) {
    final w = history.data.width;
    final h = history.data.height;

    for (final coord in _parseCoordList(abList)) {
      final xy = Coords.sgfToXY(coord, w, h);
      if (xy != null) {
        final idx = xy[0] * h + xy[1];
        history.data.stones[idx] = Stone.black;
        zobrist.toggleStone(xy[0], xy[1], w, h, Stone.black);
      }
    }
    for (final coord in _parseCoordList(awList)) {
      final xy = Coords.sgfToXY(coord, w, h);
      if (xy != null) {
        final idx = xy[0] * h + xy[1];
        history.data.stones[idx] = Stone.white;
        zobrist.toggleStone(xy[0], xy[1], w, h, Stone.white);
      }
    }
  }

  /// Parse a space-separated list of SGF coordinate strings.
  static List<String> _parseCoordList(String s) {
    if (s.trim().isEmpty) return [];
    return s.trim().split(RegExp(r'\s+'));
  }

  /// Parse the content after the root node, handling moves and `(...)` variations.
  static void _parseBranchContent(
    String content,
    BoardHistoryList history,
    int width,
    int height,
    Zobrist zobrist,
  ) {
    int i = 0;
    final nodeStack = <_NodeFrame>[];

    while (i < content.length) {
      final ch = content[i];

      if (ch == ';') {
        // Node start — find the end of this node.
        final nodeEnd = _findNodeEnd(content, i + 1);
        final nodeContent = content.substring(i + 1, nodeEnd);
        _applyNode(nodeContent, history, width, height, zobrist);
        i = nodeEnd;
      } else if (ch == '(') {
        // Start of a variation branch.
        nodeStack.add(
          _NodeFrame(node: history.head, depth: _countOpenParens(content, i)),
        );
        i++;
      } else if (ch == ')') {
        // End of a variation branch. Restore the parent node.
        if (nodeStack.isNotEmpty) {
          final frame = nodeStack.removeLast();
          // Go back to the node where the branch started.
          history.head = frame.node;
        }
        i++;
      } else {
        i++;
      }
    }
  }

  /// Find the end of a node (next ';', '(', ')', or end of string).
  static int _findNodeEnd(String s, int start) {
    bool inBracket = false;
    bool escape = false;
    for (int i = start; i < s.length; i++) {
      if (escape) {
        escape = false;
        continue;
      }
      final ch = s[i];
      if (ch == '\\') {
        escape = true;
        continue;
      }
      if (ch == '[') {
        inBracket = true;
        continue;
      }
      if (ch == ']') {
        inBracket = false;
        continue;
      }
      if (!inBracket && (ch == ';' || ch == '(' || ch == ')')) {
        return i;
      }
    }
    return s.length;
  }

  /// Extract properties from a node content string (between ; and next node/bracket).
  static void _extractProperties(String content, Map<String, String> props) {
    final propMatch = RegExp(
      r'([A-Za-z]+)((?:\[[^\]]*\])+)',
    ).allMatches(content);
    for (final m in propMatch) {
      final name = m.group(1)!;
      final values = m.group(2)!;
      // Extract ALL bracket values, space-separated (for multi-value props like AB[cd][ef]).
      final allValues = RegExp(r'\[([^\]]*)\]').allMatches(values);
      final value = allValues.map((vm) => vm.group(1) ?? '').join(' ');
      props[name] = value;
    }
  }

  /// Apply a single node's properties to the history (making a move if B or W).
  static void _applyNode(
    String nodeContent,
    BoardHistoryList history,
    int width,
    int height,
    Zobrist zobrist,
  ) {
    final props = <String, String>{};
    _extractProperties(nodeContent, props);

    final bCoord = props['B'] ?? props['b'] ?? '';
    final wCoord = props['W'] ?? props['w'] ?? '';

    if (bCoord.isNotEmpty) {
      _playFromSgf(bCoord, Stone.black, history, width, height, zobrist);
    } else if (wCoord.isNotEmpty) {
      _playFromSgf(wCoord, Stone.white, history, width, height, zobrist);
    }

    // Store comment if present.
    final comment = props['C'] ?? props['c'] ?? '';
    if (comment.isNotEmpty) {
      history.data.comment = _unescapeSgf(comment);
    }

    // Store remaining properties.
    for (final entry in props.entries) {
      if (!['B', 'b', 'W', 'w', 'C', 'c'].contains(entry.key)) {
        history.data.properties[entry.key] = entry.value;
      }
    }
  }

  /// Play a move from an SGF coordinate string.
  static void _playFromSgf(
    String coord,
    Stone color,
    BoardHistoryList history,
    int width,
    int height,
    Zobrist zobrist,
  ) {
    final xy = Coords.sgfToXY(coord, width, height);
    final stones = history.data.stones.toList(growable: false);
    final zobristClone = zobrist.clone();
    final moveNumber = history.data.moveNumber + 1;

    if (xy == null) {
      // Pass.
      final newState = BoardData(
        width: width,
        height: height,
        stones: stones,
        lastMove: null,
        lastMoveColor: color,
        blackToPlay: !history.data.blackToPlay,
        zobrist: zobristClone.hash,
        moveNumber: moveNumber,
        moveNumberList: List<int>.filled(stones.length, 0),
        blackCaptures: history.data.blackCaptures,
        whiteCaptures: history.data.whiteCaptures,
        winrate: 100.0 - history.data.winrate,
      );
      history.addOrGoto(newState, false, false);
      return;
    }

    final x = xy[0];
    final y = xy[1];
    final idx = x * height + y;
    stones[idx] = color;
    zobristClone.toggleStone(x, y, width, height, color);

    final newState = BoardData(
      width: width,
      height: height,
      stones: stones,
      lastMove: [x, y],
      lastMoveColor: color,
      blackToPlay: !history.data.blackToPlay,
      zobrist: zobristClone.hash,
      moveNumber: moveNumber,
      moveNumberList: List<int>.filled(stones.length, 0),
      blackCaptures: history.data.blackCaptures,
      whiteCaptures: history.data.whiteCaptures,
      winrate: 100.0 - history.data.winrate,
    );
    // Sync the zobrist back so the next move toggles from this state.
    zobrist.setHash(zobristClone.hash);
    history.addOrGoto(newState, false, false);
  }

  /// Count open parentheses from position [start] to the current position.
  static int _countOpenParens(String s, int start) {
    int count = 0;
    for (int i = 0; i <= start && i < s.length; i++) {
      if (s[i] == '(') count++;
      if (s[i] == ')') count--;
    }
    return count;
  }

  /// Unescape SGF text (replace \\ with \, \] with ]).
  static String _unescapeSgf(String s) {
    return s
        .replaceAll('\\\\', '\\')
        .replaceAll('\\]', ']')
        .replaceAll('\\n', '\n');
  }

  /// Escape SGF text (replace \ with \\, ] with \]).
  static String _escapeSgf(String s) {
    return s.replaceAll('\\', '\\\\').replaceAll(']', '\\]');
  }

  // ---------------------------------------------------------------------------
  // Serialization
  // ---------------------------------------------------------------------------

  /// Serialize a [BoardHistoryList] to SGF string.
  ///
  /// Optionally accepts a [GameInfo] for metadata (komi, handicap, player names).
  /// If omitted, default values are used.
  static String serialize(
    BoardHistoryList history, {
    String appName = 'Lizzie',
    GameInfo? gameInfo,
  }) {
    final buf = StringBuffer();
    final root = history.root;
    final data = root.data;
    final info = gameInfo ?? GameInfo();

    buf.write('(;GM[1]FF[4]');
    buf.write('SZ[${data.width}]');
    buf.write('KM[${info.komi}]');
    if (info.handicap > 0) {
      buf.write('HA[${info.handicap}]');
    }
    if (info.playerBlack.isNotEmpty) {
      buf.write('PB[${_escapeSgf(info.playerBlack)}]');
    }
    if (info.playerWhite.isNotEmpty) {
      buf.write('PW[${_escapeSgf(info.playerWhite)}]');
    }
    if (info.result.isNotEmpty) {
      buf.write('RE[${_escapeSgf(info.result)}]');
    }
    if (info.gameName.isNotEmpty) {
      buf.write('GN[${_escapeSgf(info.gameName)}]');
    }
    if (info.gameDate.isNotEmpty) {
      buf.write('DT[${_escapeSgf(info.gameDate)}]');
    }
    buf.write('AP[$appName]');

    // Root comment.
    if (data.comment.isNotEmpty) {
      buf.write('C[${_escapeSgf(data.comment)}]');
    }

    // Serialize move tree.
    if (root.next != null) {
      _serializeNode(buf, root, data.width, data.height);
    }

    buf.write(')');
    return buf.toString();
  }

  /// Recursively serialize a node and its variations.
  static void _serializeNode(
    StringBuffer buf,
    BoardHistoryNode node,
    int width,
    int height,
  ) {
    // Skip the root node (no move to write).
    if (node.previous != null) {
      final data = node.data;
      final lastMove = data.lastMove;
      final color = data.lastMoveColor;
      final moveStr = color == Stone.black ? 'B' : 'W';

      if (lastMove == null) {
        buf.write(';$moveStr[]');
      } else {
        final sgfCoord = Coords.xyToSgf(lastMove[0], lastMove[1]);
        buf.write(';$moveStr[$sgfCoord]');
      }

      // Comment.
      if (data.comment.isNotEmpty) {
        buf.write('C[${_escapeSgf(data.comment)}]');
      }
    }

    // Handle variations.
    if (node.childCount > 1) {
      // Main branch — first child is always present when childCount > 1.
      buf.write('(');
      _serializeNode(buf, node.variations[0], width, height);
      buf.write(')');
      // Side variations.
      for (int i = 1; i < node.variations.length; i++) {
        buf.write('(');
        _serializeNode(buf, node.variations[i], width, height);
        buf.write(')');
      }
    } else if (node.childCount == 1) {
      // Single continuation — no parentheses needed.
      _serializeNode(buf, node.variations[0], width, height);
    }
  }
}

/// Internal helper for tracking branch depth during SGF parsing.
class _NodeFrame {
  final BoardHistoryNode node;
  final int depth;

  const _NodeFrame({required this.node, required this.depth});
}
